lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+module StatsD
+  module Instrument
+    # @note This class is part of the new Client implementation that is intended
+    #   to become the new default in the next major release of this library.
+    class BatchedUDPSink
+      DEFAULT_FLUSH_INTERVAL = 1.0
+      MAX_PACKET_SIZE = 508
+
+      def self.for_addr(addr, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        host, port_as_string = addr.split(":", 2)
+        new(host, Integer(port_as_string), flush_interval: flush_interval)
+      end
+
+      attr_reader :host, :port
+
+      def initialize(host, port, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        @host = host
+        @port = port
+        @mutex = Mutex.new
+        @socket = nil
+        @flush_interval = flush_interval
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
+        @dispatcher_thread = nil
+        spawn_dispatcher
+      end
+
+      def sample?(sample_rate)
+        sample_rate == 1 || rand < sample_rate
+      end
+
+      def <<(datagram)
+        @buffer << datagram
+        spawn_dispatcher unless @dispatcher_thread&.alive?
+        self
+      end
+
+      def shutdown
+        @dispatcher_thread&.kill
+      end
+
+      private
+
+      def spawn_dispatcher
+        @dispatcher_thread = Thread.new { dispatch }
+      end
+
+      NEWLINE = "\n".b.freeze
+      def flush
+        return if @buffer.empty?
+
+        datagrams = @buffer.shift(@buffer.size)
+
+        until datagrams.empty?
+          packet = String.new(datagrams.pop, encoding: Encoding::BINARY, capacity: MAX_PACKET_SIZE)
+
+          until datagrams.empty? || packet.bytesize + datagrams.first.bytesize + 1 > MAX_PACKET_SIZE
+            packet << NEWLINE << datagrams.shift
+          end
+
+          send_packet(packet)
```

Comment:
Note that it looks like linux supports sending multiple packets in a single system call with [sendmmsg](https://linux.die.net/man/2/sendmmsg), although it doesn't look ruby provides a way to use it without a native extension

Reply:
Yeah, would be worth to try exposing it with a small C-ext. But now that `sendmsg` is in a background thread I don't think the difference will be that big.